crates/storage/src/store/substore.rs

```diff
@@ -0,0 +1,233 @@
+            cf_jmt_keys_by_keyhash: format!("substore-{}-jmt-keys-by-keyhash", prefix),
+            cf_nonverifiable: format!("substore-{}-nonverifiable", prefix),
+            prefix,
+        }
+    }
+
+    pub fn transparent_store() -> Self {
+        Self {
+            // TODO: harmonize cf format throughout.
+            cf_jmt: "jmt".to_string(),
+            cf_jmt_keys: "jmt_keys".to_string(),
+            cf_jmt_values: "jmt_values".to_string(),
+            cf_jmt_keys_by_keyhash: "jmt_keys_by_keyhash".to_string(),
+            cf_nonverifiable: "nonverifiable".to_string(),
+            prefix: "".to_string(),
+        }
+    }
+
+    /// Returns an iterator over all column families in this substore.
+    /// This is verbose, but simple.
+    pub fn columns(&self) -> impl Iterator<Item = &String> {
+        std::iter::once(&self.cf_jmt)
+            .chain(std::iter::once(&self.cf_jmt_keys))
+            .chain(std::iter::once(&self.cf_jmt_values))
+            .chain(std::iter::once(&self.cf_jmt_keys_by_keyhash))
+            .chain(std::iter::once(&self.cf_nonverifiable))
+    }
+
+    pub fn cf_jmt<'s>(&self, snapshot: &'s Snapshot) -> &'s ColumnFamily {
+        snapshot
+            .0
+            .db
+            .cf_handle(self.cf_jmt.as_str())
+            .expect("substore jmt column family not found")
+    }
+
+    pub fn cf_jmt_values<'s>(&self, snapshot: &'s Snapshot) -> &'s ColumnFamily {
+        snapshot
+            .0
+            .db
+            .cf_handle(self.cf_jmt_values.as_str())
+            .expect("substore jmt-values column family not found")
+    }
+
+    pub fn cf_jmt_keys_by_keyhash<'s>(&self, snapshot: &'s Snapshot) -> &'s ColumnFamily {
+        snapshot
+            .0
+            .db
+            .cf_handle(self.cf_jmt_keys_by_keyhash.as_str())
+            .expect("substore jmt-keys-by-keyhash column family not found")
+    }
+
+    pub fn cf_jmt_keys<'s>(&self, snapshot: &'s Snapshot) -> &'s ColumnFamily {
+        snapshot
+            .0
+            .db
+            .cf_handle(self.cf_jmt_keys.as_str())
+            .expect("substore jmt-keys column family not found")
+    }
+
+    pub fn cf_nonverifiable<'s>(&self, snapshot: &'s Snapshot) -> &'s ColumnFamily {
+        snapshot
+            .0
+            .db
+            .cf_handle(&self.cf_nonverifiable.as_str())
+            .expect("substore jmt-keys column family not found")
+    }
+    // TODO: we can use a `rocksdb::OptimisticTransactionDB` since we know that
+    // our write load is not contentious (definitionally), and we can use make
+    // writing to every substore atomic.
+    pub fn _commit(&self, _changeset: ()) -> Result<()> {
+        todo!("commit changeset to rocksdb")
+    }
+}
+
+pub struct SubstoreSnapshot {
+    pub config: Arc<SubstoreConfig>,
+    pub snapshot: Snapshot,
+}
+
+impl SubstoreSnapshot {
+    fn rocksdb_snapshot(&self) -> &RocksDbSnapshot {
+        &self.snapshot.0.snapshot
+    }
+}
```

Comment:
The existence of this method suggests that the data modeling for the `SubstoreSnapshot` is wrong, and that it would be better to have it just hold on to the `RocksDbSnapshot`. That way we don't have this ugly pass-the-parent-container-to-the-sub-object pattern of cloning the snapshot inside itself.